File: June-LeetCoding-Challenge/18-H-Index-II/H-Index-II.cpp
class Solution {
public:
    int hIndex(vector<int>& citations) {
        int n = citations.size();
        int left = 0, right = n - 1;
        while (left <= right) {
            int mid = (right - left) / 2 + left;
            if (n - mid <= citations[mid]) {
                right = mid - 1;
            }
            else left = mid + 1;
        }
        return n - left;
    }
};


// solution when the citations vector is not initially sorted.
class Solution2 {
public:
    int hIndex(vector<int>& citations) {
        make_heap(citations.begin(), citations.end(), greater<int>{});
        while(!citations.empty() && citations.front() < citations.size()){
            pop_heap(citations.begin(), citations.end(), greater<int>{});
            citations.pop_back();
        }
        
        return citations.size();
    }
};

File: June-LeetCoding-Challenge/18-H-Index-II/H-Index-II.kt
class HIndexIIKotlin275 {
    fun hIndex(citations: IntArray): Int {
        if (citations.isEmpty()) {
            return 0
        }
        var left = 0
        var right = citations.size
        while (left + 1 < right) {
            val mid = left + (right - left) / 2
            when {
                citations[citations.size - mid] >= mid -> left = mid
                else -> right = mid
            }
        }
        return when {
            citations[citations.size - right] >= right -> right
            left == 0 -> 0
            citations[citations.size - left] >= left -> left
            else -> -1
        }
    }
}